api/tests/test_graphql.rs

```diff
@@ -0,0 +1,198 @@
+use diesel::{
+    associations::HasTable, query_source::Table, PgConnection, RunQueryDsl,
+};
+use failure::Fallible;
+use gdlk_api::{
+    models,
+    server::{create_gql_schema, Context, GqlSchema},
+    util,
+};
+use juniper::{
+    graphql_value, DefaultScalarValue, ExecutionError, InputValue, Variables,
+};
+use maplit::hashmap;
+use std::{collections::HashMap, sync::Arc};
+
+/// Macro to run a DELETE statement for each given table. Used to clean up all
+/// tables after each test.
+macro_rules! delete_tables {
+    ($conn:expr, $($model:ty),+ $(,)?) => {
+        $(
+            diesel::delete(<$model>::table())
+            .execute($conn as &PgConnection)
+            .unwrap();
+        )+
+    };
+}
+
+/// Helper type for setting up and executing test GraphQL queries
+struct QueryRunner {
+    schema: GqlSchema,
+    context: Context,
+}
+
+impl QueryRunner {
+    fn new() -> Fallible<Self> {
+        let pool = util::init_db_conn_pool()?;
+
+        Ok(Self {
+            schema: create_gql_schema(),
+            context: Context {
+                pool: Arc::new(pool),
+            },
+        })
+    }
+
+    fn query<'a>(
+        &'a self,
+        query: &'a str,
+        vars: HashMap<&str, InputValue>,
+    ) -> (
+        juniper::Value<DefaultScalarValue>,
+        Vec<ExecutionError<DefaultScalarValue>>,
+    ) {
+        // Map &strs to Strings
+        let converted_vars = vars
+            .into_iter()
+            .map(|(k, v)| (k.to_string(), v))
+            .collect::<Variables>();
+
+        juniper::execute(
+            query,
+            None,
+            &self.schema,
+            &converted_vars,
+            &self.context,
+        )
+        .unwrap()
+    }
+}
+
+// Automatically wipe all tables when the test is done
+impl Drop for QueryRunner {
+    fn drop(&mut self) {
+        let conn = self.context.get_db_conn().unwrap();
+        delete_tables!(
+            &conn,
+            models::User,
+            models::HardwareSpec,
+            models::ProgramSpec,
+            models::UserProgram,
+            // Any new table needs to be added here!
+        );
```

Comment:
We need a way to have the tests not affect each other via the DB. I tried transactions but that doesn't work because the API uses a connection pool and each transaction is attached to one connection. This isn't a great solution since we have to maintain this list, but it's the best I can come up with.

Reply:
Could we limit the connection pool for the tests so transactions work?

Reply:
Still this drop trait is so lit that it just happen automagically